File: backend/enum-jpa-validation-app/src/main/java/com/example/enumapp/web/validation/NoNullElements.java
package com.example.enumapp.web.validation;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/** List 자체는 허용하되, 요소 중 null 이 있으면 실패. */
@Documented
@Target({ElementType.FIELD, ElementType.PARAMETER})
@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy = NoNullElementsValidator.class)
public @interface NoNullElements {

    String message() default "list must not contain null elements";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};
}

File: backend/enum-jpa-querydsl-kotlin-app/src/main/kotlin/com/example/enumapp/web/validation/NoNullElements.kt
package com.example.enumapp.web.validation

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.annotation.AnnotationRetention.RUNTIME
import kotlin.annotation.AnnotationTarget.FIELD
import kotlin.annotation.AnnotationTarget.VALUE_PARAMETER
import kotlin.reflect.KClass

/** List 자체는 허용하되, 요소 중 null 이 있으면 실패. */
@MustBeDocumented
@Target(FIELD, VALUE_PARAMETER)
@Retention(RUNTIME)
@Constraint(validatedBy = [NoNullElementsValidator::class])
annotation class NoNullElements(
    val message: String = "list must not contain null elements",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = [],
)
